### PR TITLE
Use order date for sorting and display if no error

### PR DESCRIPTION
**Pull request by Rad1calp**
Displays the date the order was created if there is not an error attached to the failure.
Sorting now uses the order date if there is no error attached to the failure.
Sorting now is newest first.

### CODE REVIEW

File: admin/failures/overview/endpoint.js
Comment (hypesystem) [inline admin/failures/overview/endpoint.js:19]: This is some of the most difficult-to-read code I've ever seen :smile:

Keep it simple: do simple if-statements, extract variables, etc. instead of this. Readability is more important than almost anything else!

File: admin/failures/overview/endpoint.js
Comment (hypesystem) [inline admin/failures/overview/endpoint.js:33]: Extract `latestErrorAt` to a variable right after `const latestError`:

```js
const latestErrorAt = latestError ? prettifyDate(latestError.at) : prettifyDate(o.data.viewModel.date);
```

Then this line can be:

```js
latestErrorAt,
```

Which makes the whole thing more readable

File: admin/failures/overview/endpoint.js
Comment (Rad1calp) [reply admin/failures/overview/endpoint.js]: should be fixed now

File: admin/failures/overview/endpoint.js
Comment (hypesystem) [inline admin/failures/overview/endpoint.js:28]: Having a pattern like this, where you do the same thing on two items is a perfect case for extracting a function!

So `sort` becomes:

```
.sort((a, b) => {
  const aLatestTimestamp = getLatestTimestamp(a);
  const bLatestTimestamp = getLatestTimestamp(b);

  // ...
});
```

File: admin/failures/overview/endpoint.js
Comment (hypesystem) [inline admin/failures/overview/endpoint.js:39]: This would extract nicely into a function `getLastElement(o.data.errors)`, in which you spell out the if-statements and return early to make the code super clear:

```js
function getLastELement(list) {
  if(!list || !list.length) {
    return undefined;
  }
  return list[list.length - 1];
}
```

See how easy that is to skim and understand??  :smile:

File: admin/failures/overview/endpoint.js
Comment (hypesystem) [inline admin/failures/overview/endpoint.js:40]: Good case for an extraction here, too. You can even do a nice trick by doing a general function:

```js
getFirstPrettyDate(latestError?.at, o.data.viewModel.date);

// where...

function getFirstPrettyDate(...dateStrings) {
    for(const dateString of dateStrings) {
        if(dateString) {
            return prettyDate(dateString);
        }
    }
}
```

Notice the rest parameters syntax (`...dateStrings`): https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters

Comment (hypesystem) [reply admin/failures/overview/endpoint.js:40]: In general, a function should do one thing on one level of abstraction.

In this case we have two levels of abstraction mixed:

- Deciding where to get a date from, based on which data exists (relatively high-level business logic decision)
- Transforming a date into the correct display format

When it is in a simple teritary we mix the two decisions. My suggestion is that we create a general function that returns a date from several prioritized possibilities (*without* knowing the shape or purpose of our data!) and a higher level of abstraction making the decision on data and simply passing it to a function.

This splits the two levels of abstraction and makes it easier to reason about.

**Review by hypesystem — APPROVED** [review]
Generally very good and an improvement!

Added some few more notes that I think you should incorporate, but then you are welcome to merge.

In general, don't be afraid of extracting functions - they are basically free and give you a neat label to describe your logical intent!

Try to use function extraction to split levels of abstraction neatly, so each bit of code can be read without traversing several levels of abstraction.

File: admin/failures/overview/endpoint.js
Comment (hypesystem) [inline admin/failures/overview/endpoint.js:53]: Now that you have a `getLastElement` helper function, you can use it here to make the code even simpler:

```js
const lastError = getLastElement(order.data.errors);
if(lastError) {
    return lastError.at;
}
return order.data.viewModel.date;
```

This is *more lines* but conceptually simpler and easier to read because we separate the two levels of abstraction:

- mechanics of getting last item of a list that may or may not be defined
- presentational logic of getting a date or doing a fallback to a view model date

File: admin/failures/overview/endpoint.js
Comment (hypesystem) [inline admin/failures/overview/endpoint.js:32]: Instead of getting `latestError` and `latestErrorAt` you can now just use your nice function `getLatestTimestamp`:

```js
latestErrorAt: prettifyDate(getLatestTimestamp(o)),
```

**Review by hypesystem — APPROVED** [review]
Generally looks very good! Now that you've taken some refactoring steps, some new ones appear! Let's take those last two ones, too - reusing getLastElement and getLatestTimestamp where they are also used.

Once you've done that feel free to merge - you don't need another round of comments from me

**Comment by hypesystem** [comment]
This is already approved - and still is after that last commit :smile:

**Pull request merged**
